cpp/src/parquet/arrow/reader.cc

```diff
@@ -547,11 +558,19 @@ Status FileReader::Impl::ReadRowGroup(int row_group_index,
                          this](int i) {
     std::shared_ptr<ChunkedArray> array;
     RETURN_NOT_OK(ReadColumnChunk(field_indices[i], indices, row_group_index, &array));
+
+    // Update the schema when reading dictionary array
+    if (reader_properties_.read_dictionary(i)) {
+      auto dict_field =
+          std::make_shared<::arrow::Field>(schema->field(i)->name(), array->type());
+      RETURN_NOT_OK(schema->SetField(i, dict_field, &schema));
```

Comment:
This fails to merge dictionaries across multiple row groups.  I think I can use `arrow::DictionaryType::Unify` to solve this.  Working on fixing this and adding tests.

Reply:
Might be better to "fix" the schema all at once rather than having to call `SetField` possibly multiple times. That can be done later...